File: oc/UIBasicExam/UIBasicExam/exam5/Exam5ViewController.m
//
//  Exam5ViewController.m
//  UIBasicExam
//
//  Created by dfang on 2017-12-25.
//  Copyright © 2017年 com.east. All rights reserved.
//

#import "Exam5ViewController.h"

@interface Exam5ViewController () <UIScrollViewDelegate>

@property (nonatomic, strong) UIScrollView *scrollView;
@property (nonatomic, strong) UIImageView *sunImageView;
@property (nonatomic, strong) UIImageView *contentImageView;

@end

@implementation Exam5ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.view.backgroundColor = [UIColor whiteColor];
    
    CGSize screenSize = [UIScreen mainScreen].bounds.size;
    _scrollView = [[UIScrollView alloc] initWithFrame:CGRectMake(0, 0, screenSize.width, screenSize.height)];
    _scrollView.delegate = self;
    [self.view addSubview:_scrollView];
    
    UIImageView *bg = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"520_userguid_bg"]];
    [_scrollView addSubview:bg];
    
    UIImageView *cloud = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"520_userguid_cloud"]];
    [_scrollView addSubview:cloud];
    
    UIImageView *bottom = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"520_userguid_fg"]];
    bottom.frame = CGRectMake(0, screenSize.height - bottom.frame.size.height, bottom.frame.size.width, bottom.frame.size.height);
    [_scrollView addSubview:bottom];
    
    UIImageView *words = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"520_userguid_words_5"]];
    words.frame = CGRectMake(0, 70, screenSize.width, 100);
    [_scrollView addSubview:words];
    
    UIImageView *cg = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"520_userguid_cg"]];
    [_scrollView addSubview:cg];
    
    _sunImageView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"520_userguid_sun"]];
    _sunImageView.frame = CGRectMake(screenSize.width - 60, 10, 60, 60);
    [_scrollView addSubview:_sunImageView];
    
    _contentImageView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"520_userguid_person_weigou"]];
    _contentImageView.frame = CGRectMake((screenSize.width - 150)/2, screenSize.height - 200, 150, 150);
    NSArray *moveArray = @[[UIImage imageNamed:@"520_userguid_person_taitou_1"], [UIImage imageNamed:@"520_userguid_person_ditou_1"], [UIImage imageNamed:@"520_userguid_person_taitou_2"], [UIImage imageNamed:@"520_userguid_person_ditou_2"]];
    _contentImageView.animationImages = moveArray;
    _contentImageView.animationDuration = 0.5;
    [_scrollView addSubview:_contentImageView];
    
    _scrollView.contentSize = CGSizeMake(CGRectGetMaxX(bg.frame), screenSize.height);
}

- (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView {
    [_contentImageView startAnimating];
    
    CABasicAnimation* rotationAnimation;
    rotationAnimation = [CABasicAnimation animationWithKeyPath:@"transform.rotation.z"];
    rotationAnimation.toValue = [NSNumber numberWithFloat: M_PI * 2.0 ];
    rotationAnimation.duration = 1;
    rotationAnimation.cumulative = YES;
    rotationAnimation.repeatCount = 100;
    // 开始动画
    [_sunImageView.layer addAnimation:rotationAnimation forKey:@"rotationAnimation"];
}

- (void)scrollViewDidEndDragging:(UIScrollView *)scrollView willDecelerate:(BOOL)decelerate {
    [_contentImageView stopAnimating];
    // 结束动画
    [_sunImageView.layer removeAnimationForKey:@"rotationAnimation"];
}

- (void)scrollViewDidScroll:(UIScrollView *)scrollView {
    CGFloat moveX = scrollView.contentOffset.x;
    CGRect sumFrame = _sunImageView.frame;
    CGRect contentFrame = _contentImageView.frame;
    sumFrame.origin.x = moveX + (scrollView.frame.size.width - 60);
    contentFrame.origin.x = moveX + (scrollView.frame.size.width - 150)/2;
    
    _sunImageView.frame = sumFrame;
    _contentImageView.frame = contentFrame;
}

@end

File: swift/UIBasicExam/UIBasicExam/exam5/Exam5ViewController.swift
//
//  Exam5ViewController.swift
//  UIBasicExam
//
//  Created by dfang on 2017-12-28.
//  Copyright © 2017年 dfang. All rights reserved.
//

import Foundation
import UIKit

class Exam5ViewController: UIViewController, UIScrollViewDelegate {
    let scrollView = UIScrollView.init()
    let sunImageView = UIImageView.init()
    let moveImageView = UIImageView.init()
    
    override func viewDidLoad() {
        scrollView.frame = self.view.frame
        scrollView.delegate = self
        self.view.addSubview(scrollView)
        
        let bgImageView = UIImageView.init(image: UIImage.init(named: "520_userguid_bg"))
        scrollView.addSubview(bgImageView)
        
        let cloudImageView = UIImageView.init(image: UIImage.init(named: "520_userguid_cloud"))
        scrollView.addSubview(cloudImageView)
        
        let footerBgImageView = UIImageView.init(image: UIImage.init(named: "520_userguid_fg"))
        var origin = footerBgImageView.frame.origin
        origin.y = screenHeight - footerBgImageView.frame.size.height
        footerBgImageView.frame.origin = origin
        scrollView.addSubview(footerBgImageView)
        
        let wordsImageView = UIImageView.init(image: UIImage.init(named: "520_userguid_words_5"))
        wordsImageView.frame = CGRect.init(x: 0, y: 70, width: screenWidth, height: 100)
        scrollView.addSubview(wordsImageView)
        
        let sceneImageView = UIImageView.init(image: UIImage.init(named: "520_userguid_cg"))
        scrollView.addSubview(sceneImageView)
        
        sunImageView.frame = CGRect.init(x: screenWidth - 60, y: 10, width: 60, height: 60)
        sunImageView.image = UIImage.init(named: "520_userguid_sun")
        scrollView.addSubview(sunImageView)
        
        moveImageView.frame = CGRect.init(x: (screenWidth - 150)/2, y: screenHeight - 200, width: 150, height: 200)
        moveImageView.image = UIImage.init(named: "520_userguid_person_weigou")
        moveImageView.animationImages = [UIImage.init(named: "520_userguid_person_taitou_1")!,
                                         UIImage.init(named: "520_userguid_person_ditou_1")!,
                                         UIImage.init(named: "520_userguid_person_taitou_2")!,
                                         UIImage.init(named: "520_userguid_person_ditou_2")!]
        moveImageView.animationDuration = 0.5
        scrollView.addSubview(moveImageView)
        
        scrollView.contentSize = CGSize.init(width: bgImageView.frame.maxX, height: screenHeight)
    }
    
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        let moveX = scrollView.contentOffset.x
        var sunOrigin = sunImageView.frame.origin
        sunOrigin.x = moveX + (scrollView.frame.width - 60)
        sunImageView.frame.origin = sunOrigin
        
        var moveOrigin = moveImageView.frame.origin
        moveOrigin.x = moveX + (scrollView.frame.width - 150)/2
        moveImageView.frame.origin = moveOrigin
    }
    
    func scrollViewWillBeginDragging(_ scrollView: UIScrollView) {
        moveImageView.startAnimating()
        
        let rotationAnimation = CABasicAnimation.init(keyPath: "transform.rotation.z")
        rotationAnimation.fromValue = NSNumber.init(value: 0)
        rotationAnimation.toValue = NSNumber.init(value: Double.pi * 2.0)
        rotationAnimation.duration = 1
        rotationAnimation.repeatCount = 100
        sunImageView.layer.add(rotationAnimation, forKey: "rotationAnimation")
        
    }
    
    func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
        moveImageView.stopAnimating()
        
        sunImageView.layer.removeAllAnimations()
    }
}
